java/fury-core/src/main/java/org/apache/fury/io/FuryReadableChannel.java

```diff
@@ -20,30 +20,143 @@
 package org.apache.fury.io;
 
 import java.io.IOException;
-import java.io.UnsupportedEncodingException;
 import java.nio.ByteBuffer;
 import java.nio.channels.ReadableByteChannel;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.fury.exception.DeserializationException;
 import org.apache.fury.memory.MemoryBuffer;
+import org.apache.fury.util.Platform;
 
-// TODO support zero-copy channel reading.
-public class FuryReadableChannel extends AbstractStreamReader implements ReadableByteChannel {
+@NotThreadSafe
+public class FuryReadableChannel implements FuryStreamReader, ReadableByteChannel {
   private final ReadableByteChannel channel;
-  private final ByteBuffer byteBuffer;
-  private final MemoryBuffer buffer;
+  private final MemoryBuffer memoryBuffer;
+  private ByteBuffer byteBuffer;
 
   public FuryReadableChannel(ReadableByteChannel channel) {
     this(channel, ByteBuffer.allocate(4096));
   }
 
-  private FuryReadableChannel(ReadableByteChannel channel, ByteBuffer directBuffer) {
+  public FuryReadableChannel(ReadableByteChannel channel, ByteBuffer directBuffer) {
     this.channel = channel;
     this.byteBuffer = directBuffer;
-    this.buffer = MemoryBuffer.fromByteBuffer(directBuffer);
+
+    int offset = directBuffer.arrayOffset() + directBuffer.position();
```

Comment:
We need an offheap buffer here. Otherwise wr can just warp channel into a fury input stream